worker/transfer.go

```diff
@@ -123,68 +126,129 @@ func parallelUploadSlab(ctx context.Context, sp storeProvider, shards [][]byte,
 		span.End()
 	}
 
-	// spawn workers and send initial requests
-	hostIndex := 0
+	// helper to track used hosts.
+	hostUsed := make([]bool, len(contracts))
+	nextHost := func() int {
+		for i := range hostUsed {
+			if !hostUsed[i] {
+				hostUsed[i] = true
+				return i
+			}
+		}
+		return -1
+	}
+
+	// helper to determine whether a request is finished.
+	isReqFinished := func(r req) bool {
+		select {
+		case <-r.finishedCtx.Done():
+			return true
+		default:
+			return false
+		}
+	}
+
+	// helper function for launching worker given an existing request. Returns
+	// true if a worker was launched for the request or if the request was
+	// already finished.
 	inflight := 0
-	for i := range shards {
-		go worker(req{contracts[hostIndex], i})
-		hostIndex++
+	launchWorker := func(r req) bool {
+		if isReqFinished(r) {
+			return true
+		}
+		hostIndex := nextHost()
+		if hostIndex == -1 {
+			return false
+		}
+		go worker(req{r.finishedCtx, r.finishedFn, hostIndex, r.shardIndex})
 		inflight++
+		return true
+	}
+
+	// spawn workers and send initial requests
+	for i := range shards {
+		finishedCtx, finishedFn := context.WithCancel(ctx)
+		if !launchWorker(req{finishedCtx, finishedFn, -1, i}) {
+			panic("failed to launch worker for initial shard - should never happen")
+		}
 	}
 
 	// collect responses
 	var errs HostErrorSet
 	sectors := make([]object.Sector, len(shards))
 	rem := len(shards)
-	for rem > 0 && inflight > 0 {
+	var toOverdrive []req
+	var slowRequests []req
+	for inflight > 0 {
 		resp := <-respChan
 		if !errors.Is(resp.err, errUploadSectorTimeout) {
 			inflight--
 		}
 
-		if resp.err != nil {
-			errs = append(errs, &HostError{resp.req.contract.HostKey, resp.err})
-			// try next host
-			if hostIndex < len(contracts) {
-				go worker(req{contracts[hostIndex], resp.req.shardIndex})
-				hostIndex++
-				inflight++
+		// Decide whether to reuse the host or not. We only do that if the
+		// worker finished executing and failed due to the request being
+		// finished already.
+		if resp.err != nil && !errors.Is(resp.err, errUploadSectorTimeout) {
+			// Remember the error and reuse the host if the request
+			// was already finished.
+			errs = append(errs, &HostError{contracts[resp.req.hostIndex].HostKey, resp.err})
```

Comment:
Why do we stop collecting timeout errors in `errs`? Seems like if inflight eventually is 0 and remaining is greater than zero we probably want to log all upload sector timeouts if we log `errs`, right? It's not that important but I don't see the reason not to append any error we get to `errs`

Reply:
Then we can also get rid of `slowRequests` again and go back to filtering the slow host indices from the errors list? I think `slowRequests` don't even make that much sense since we should probably just immediately build a `slowHosts` map. We don't use any of the `req` info other than the host index anyway.